pkg/dyninst/dwarf/loclist/loclist.go

```diff
@@ -0,0 +1,366 @@
+			}
+			op = ir.Register{Idx: uint8(idx), Shift: shift}
+
+		case opcode == dw_op_call_frame_cfa:
+			op = ir.Cfa{Offset: 0}
+
+		case dw_eval_op_lo <= opcode && opcode <= dw_eval_op_hi:
+			return nil, fmt.Errorf("unsupported stack-evaluation opcode: 0x%x", opcode)
+
+		case dw_user_op_lo <= opcode && opcode <= dw_user_op_hi:
+			return nil, fmt.Errorf("unsupported user-defined opcode: 0x%x", opcode)
+		default:
+			return nil, fmt.Errorf("unknown opcode: 0x%x", opcode)
+		}
+	}
+
+	// Single op instruction list has an implicit piece.
+	if instCnt == 1 {
+		pieces = append(pieces, ir.Piece{Size: totalByteSize, Op: op})
+		op = nil
+	}
+
+	if op != nil {
+		return nil, fmt.Errorf("unconsumed last op: %v", op)
+	}
+
+	return pieces, nil
+}
+
+func readFixedSize(reader *bytes.Buffer, size uint8) (uint64, error) {
+	switch size {
+	case 1:
+		val, err := reader.ReadByte()
+		return uint64(val), err
+	case 2:
+		return uint64(binary.LittleEndian.Uint16(reader.Next(2))), nil
+	case 4:
+		return uint64(binary.LittleEndian.Uint32(reader.Next(4))), nil
+	case 8:
+		return binary.LittleEndian.Uint64(reader.Next(8)), nil
+	default:
+		return 0, fmt.Errorf("unsupported reading %d-byte integer", size)
+	}
+}
+
+func readULEB128(reader *bytes.Buffer) (uint64, error) {
+	var val uint64
+	var shift uint
+	for {
+		b, err := reader.ReadByte()
+		if err != nil {
+			return 0, err
+		}
+		val |= uint64(b&0x7f) << shift
+		shift += 7
+		if b&0x80 == 0 {
+			break
+		}
+	}
+	return val, nil
+}
+
+func readSLEB128(reader *bytes.Buffer) (int64, error) {
+	var val int64
+	var shift uint
+	for {
+		b, err := reader.ReadByte()
+		if err != nil {
+			return 0, err
+		}
+		val |= int64(b&0x7f) << shift
+		shift += 7
+		if b&0x80 == 0 {
+			break
+		}
+	}
+	if val&0x40 != 0 {
+		val -= 1 << shift
+	}
+	return val, nil
+}
+
+// FixLoclists mitigates bug in go that can cause incorrect location lists to be generated.
+// The bug occurs because of some behavior in ssa where different pieces of a
+// variable can be interpreted as different types and can end up having
+// duplicated location lists. Empirically this has primarily been observed with
+// contexts and when it occurs, the location lists are doubled. Below is a
+// somewhat crude heuristic to detect this case and fix it.
+//
+// See https://github.com/golang/go/issues/61700
+func FixLoclists(loclists []ir.Location, expectedByteSize uint64) ([]ir.Location, error) {
+	currentByteSize := func(loclists []ir.Location) uint64 {
+		if len(loclists) == 0 || len(loclists[0].Pieces) == 0 {
+			return 0
+		}
+		var sum uint64
+		for _, piece := range loclists[0].Pieces {
+			sum += uint64(piece.Size)
+		}
+		return sum
+	}
+
+	// We want to find any pair of adjacent pieces that are in the same location
+	for currentByteSize(loclists) > expectedByteSize {
+		if len(loclists) == 0 || len(loclists[0].Pieces) == 0 {
+			break
+		}
+
+		numPieces := len(loclists[0].Pieces)
+		var foundIndex *int
```

Comment:
nit: I recognize that these indexes are small and go has that clever little hack to avoid allocating small integers, but I still think it'd be more idiomatic to have a boolean and make this just a regular old `int`.